src/java/org/apache/cassandra/service/accord/repair/AccordRepair.java

```diff
@@ -31,65 +32,93 @@
 import accord.primitives.Seekables;
 import org.apache.cassandra.dht.AccordSplitter;
 import org.apache.cassandra.dht.IPartitioner;
+import org.apache.cassandra.dht.Range;
+import org.apache.cassandra.dht.Token;
 import org.apache.cassandra.service.accord.AccordService;
+import org.apache.cassandra.service.accord.AccordTopology;
 import org.apache.cassandra.service.accord.TokenRange;
-import org.apache.cassandra.service.consensus.migration.ConsensusMigrationRepairResult;
 import org.apache.cassandra.tcm.ClusterMetadata;
 import org.apache.cassandra.tcm.Epoch;
+import org.apache.cassandra.utils.concurrent.AsyncPromise;
+import org.apache.cassandra.utils.concurrent.Future;
 
 import static com.google.common.base.Preconditions.checkState;
-import static java.util.Collections.emptyList;
 import static org.apache.cassandra.config.CassandraRelevantProperties.ACCORD_REPAIR_RANGE_STEP_UPDATE_INTERVAL;
 
 /*
  * Accord repair consists of creating a barrier transaction for all the ranges which ensure that all Accord transactions
  * before the Epoch and point in time at which the repair started have their side effects visible to Paxos and regular quorum reads.
  */
-public class AccordRepairJob extends AbstractRepairJob
+public class AccordRepair
 {
-    private static final Logger logger = LoggerFactory.getLogger(AccordRepairJob.class);
+    private static final Logger logger = LoggerFactory.getLogger(AccordRepair.class);
 
     public static final BigInteger TWO = BigInteger.valueOf(2);
 
+    interface Listener
+    {
+        void onBarrierStart();
+        void onBarrierException(Throwable throwable);
+        void onBarrierComplete();
+
+        Listener NOOP = new Listener()
+        {
+            @Override public void onBarrierStart() {}
+
+            @Override public void onBarrierException(Throwable throwable) {}
+
+            @Override public void onBarrierComplete() {}
+        };
+    }
+
+    private final Listener listener;
+
     private final Ranges ranges;
 
     private final AccordSplitter splitter;
 
     private BigInteger rangeStep;
 
-    private Epoch minEpoch = ClusterMetadata.current().epoch;
+    private final Epoch minEpoch = ClusterMetadata.current().epoch;
 
     private volatile Throwable shouldAbort = null;
 
-    public AccordRepairJob(RepairSession repairSession, String cfname)
+    public AccordRepair(Listener listener, IPartitioner partitioner, String keyspace, Collection<Range<Token>> ranges)
     {
-        super(repairSession, cfname);
-        IPartitioner partitioner = desc.ranges.iterator().next().left.getPartitioner();
-        this.ranges = AccordTopology.toAccordRanges(desc.keyspace, desc.ranges);
-        this.splitter = partitioner.accordSplitter().apply(ranges);
+        this.listener = listener != null ? listener : Listener.NOOP;
```

Comment:
`listener` is currently null, we do the following

```
AccordRepair repair = new AccordRepair(null, partitioner, desc.keyspace, desc.ranges);
```

This logic dropped metrics in favor of listeners, but doesn't have the listeners do anything yet...